### PR TITLE
Multiple stream hwbuffers

### PR DESCRIPTION
**Pull request by jeffsetter**
Modifies the internals to use multiple streams.

### CODE REVIEW

**Comment by jeffsetter** [comment]
@dillonhuff @joyliu37 
I am finally merging in the new interface to the unified buffer. The different apps and tests pass locally, but there are some differences in how the travis build operates. It would be great if you could help me debug what is going wrong. 

I also expect that some changes will be needed in other repos to accommodate the new interface.

**Comment by dillonhuff** [comment]
@jeffsetter great!

It looks as though the first problem on travis is that the hw unit tests are failing. Have you run:

```
./test/scripts/hw_unit_tests.sh 
```

on your local machine, and if so did the tests pass?

**Comment by jeffsetter** [comment]
Running that command I get:
```
At position: 4095
n_valids = 4096
n_valids = 4096
all-tests: ubuffer_tests.cpp:325: void ubuffer_small_conv_3_3_test(): Assertion `n_valids == outTileSize*outTileSize && "Valid count is not matched"' failed.
./test/scripts/run_hw_unit_tests.sh: line 27: 15408 Aborted                 (core dumped) ./all-tests
```
Any idea what's going on there? Also, in the travis build, you can see that there are issues with building BufferMapping with the coreir branch that GarnetFlow uses. Any ideas there?

**Comment by joyliu37** [comment]
> Running that command I get:
> 
> ```
> At position: 4095
> n_valids = 4096
> n_valids = 4096
> all-tests: ubuffer_tests.cpp:325: void ubuffer_small_conv_3_3_test(): Assertion `n_valids == outTileSize*outTileSize && "Valid count is not matched"' failed.
> ./test/scripts/run_hw_unit_tests.sh: line 27: 15408 Aborted                 (core dumped) ./all-tests
> ```
> 
> Any idea what's going on there? Also, in the travis build, you can see that there are issues with building BufferMapping with the coreir branch that GarnetFlow uses. Any ideas there?

@jeffsetter This is the coreIR interpreter testing after rewrite rule. I will take a closer look tomorrow. My guess is that the new compiler may not get the correct valid signal wiring or buffer parameter. And have 1 cycle valid mismatch.

**Comment by joyliu37** [comment]
I take a look at the newly extracted hardware buffer after `extract_xcel`, there are two buffer extracted. One is the input buffer which will be the line buffer. The other is the output buffer which is just a single pass through. However, the input stream of input buffer is empty, so as to the output stream of output buffer. I do not if this correct. I think it may save the access pattern of in and out?

File: src/ExtractHWBuffers.cpp [hunk 1]
```diff
@@ -1087,8 +1087,29 @@ void set_output_params(HWXcel *xcel,
...
+
+      ostream.stride_map = fos.stride_map;
+      ostream.oports = fos.oports;
+
+      ostream.odims.resize(hwbuffer.dims.size());
+      internal_assert(hwbuffer.dims.size() == ostream.odims.size());
+      for (size_t dim=0; dim<hwbuffer.dims.size(); dim++) {
+        ostream.odims[dim].output_block = hwbuffer.dims.at(dim).output_block;
+        ostream.odims[dim].output_stencil = hwbuffer.dims.at(dim).output_stencil;
+      }
```
Comment (joyliu37) [inline src/ExtractHWBuffers.cpp]: Are we suppose to add output_min/max_pos to this struct? Since it's used in the further steps.

**Comment by joyliu37** [comment]
With adding a linearized logical size(remove by false), I could get the correct rewrite result. However, it broke in the further `insert_hwbuffer` steps. It's because we have the `output_min_pos` undefined in the output stream you added last night. The code is broken at [https://github.com/StanfordAHA/Halide-to-Hardware/blob/ed0505f42e9402494ed23553a2387bb268edd469/src/InsertHWBuffers.cpp#L146](url) because it's undefined.

File: src/Lower.cpp [hunk 2]
```diff
@@ -222,70 +214,67 @@ Module lower(const vector<Function> &output_funcs, const string &pipeline_name,
...
     debug(2) << "Lowering after uniquifying variable names:\n" << s << "\n\n";
 
-    if (t.has_feature(Target::UseExtractHWKernel) && t.has_feature(Target::CoreIR)) {
-      vector<HWXcel> buf_xcels =
-        extract_hw_accelerators(s, env, inlined_stages);
-      synthesize_hwbuffers(s, env, buf_xcels);
-    }
-
+    //bool use_ubuffer = !t.has_feature(Target::UseExtractHWKernel);
+    bool use_ubuffer = true;
```
Comment (joyliu37) [inline src/Lower.cpp:218]: Previously, this flag is set to be `false`. Simply change it to `true` will break Dillon's merged unit test. It seems we are using different passes and should come up with a clear control flow to apply what in which stage. 
When I change it to false all the unit test pass.

**Comment by joyliu37** [comment]
@Kuree @rdaly525 
Jeff is creating a new coreIR unified buffer [node](https://github.com/joyliu37/BufferMapping/blob/5e6cd90f9c5af08887e60582f1724bf5b7cdbafd/cfunc/src/lakelib.cpp#L944) to support multiple input/output streams. This is a major API change, I don't know whether the downstream mapping will support this change. 
The buffermapping is currently broken since the parser is working for the old coreIR node. I am currently update that. Just want to inform you guys and get some feedbacks!

**Comment by Kuree** [comment]
@joyliu37 @jeffsetter Can you give me some examples of mapped JSON files?

**Comment by joyliu37** [comment]
Are we able to merge this branch to master? @jeffsetter Do you know what's the current error with travis build?

**Pull request closed without merging**